### PR TITLE
script: add simple signature support (checker/creator)

### PR DESCRIPTION
**Pull request by kallewoof**
The simple signature takes a sighash as argument and verifies a signature and pubkey against it.

It is used in signet (#16411) for verifying blocks and in BIP-322 implementation (#16440) for verifying messages.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
ACK 02a1db1607ec59b6c54072cd3ffb606422eea3a4

**Comment by Sjors** [comment]
Concept ACK. It strictly adds new functions so that seems a pretty safe approach. This PR needs tests though.

**Comment by emilengler** [comment]
Concept NACK until it is being used actively

**Comment by kallewoof** [comment]
@emilengler Did you read the PR description?

**Comment by kallewoof** [comment]
@Sjors Added tests. The multisig ones are based on the `multisig_tests.cpp` file, but adapted to use the new creators/checkers.

**Comment by emilengler** [comment]
@kallewoof I did but until nothing is merged I don't think this should be added.
I prefer a chronological order.

**Comment by kallewoof** [comment]
lol good trolling there

**Comment by Sjors** [comment]
@kallewoof thanks for adding the tests. They pass for me on macOS. I haven't followed the BIP-322 discussion in great detail, so can't really comment on the cryptography itself or how likely that is to change. It would be nice if it remained consistent with signet.

@emilengler when you built the binary, how many bytes were added? Even if this difference is non-trivial, it only matters if for some reason this PR makes it into the the 0.19 release and neither of the two followup PRs make the cut-off. This PR seems well worth the reduced review burden and the improved coordination between both projects.

**Comment by NicolasDorier** [comment]
On my side I am fine both about either having it in this PR or on the same PR as signet.

**Comment by emilengler** [comment]
@Sjors Sure thing but why isn't this being added to the original PR? @kallewoof

**Comment by sdaftuar** [comment]
This interface, where the caller provides the hash of a message, and then "verifies" that the signature is valid, seems problematic to me at this layer of our code -- my understanding is that without knowing the preimage of the hash (ie the message being signed), this doesn't prove that the signature could only have come from someone with knowledge of the private key.

I know at some layer of our cryptography code we will have verification methods like this, but at this layer it seems inconsistent with all our other signature checkers, which are operating under different assumptions (well, except maybe for the DummySignatureChecker which does not do signature verifications, but hopefully that is clear to any callers).  So my initial thought is that this is not a good place to have what is effectively a lower-level API call to a crypto library function.

I'm not very familiar with the broader work this is part of, so if this is in fact necessary or expedient from an implementation perspective, then I'd say at a minimum that we should at least include comments explaining that the assumptions here are different than for the other signature checkers.

**Comment by kallewoof** [comment]
@sdaftuar 

Maybe I'm misunderstanding, but I'm not sure if it's that inconsistent in the end. The "CreateSig" methods right now
1. Get the private key from the signing provider.
2. Generate the sighash based on internal data.
3. Sign the sighash using the aforementioned private key.

This is consistent across all of these, and in fact, the other implementations could become sub-classes of this one (perform step 2 into internal sighash then have master do the rest).

To pass the preimage rather than the hash itself would only lock this down to using some specific type of hasher, and would have no practical meaning for the actual implementation. The two (current) uses of these are here (#16411):

https://github.com/bitcoin/bitcoin/blob/55d7c4ef3ccf9ad84d62592969de35c4ab84af2c/src/signet.cpp#L31-L36

which gets the hash from

https://github.com/bitcoin/bitcoin/blob/55d7c4ef3ccf9ad84d62592969de35c4ab84af2c/src/signet.cpp#L54-L58

and here (#16440):

https://github.com/bitcoin/bitcoin/blob/203464efb4302d4c21bf2bc1503487f116c8837e/src/script/proof.h#L226-L233

which gets the hash from

https://github.com/bitcoin/bitcoin/blob/203464efb4302d4c21bf2bc1503487f116c8837e/src/script/proof.cpp#L15-L18

File: src/script/interpreter.h [hunk 1]
```diff
@@ -162,6 +162,18 @@ class BaseSignatureChecker
     virtual ~BaseSignatureChecker() {}
 };
 
+/** A general purpose signature checker. Note that this lacks assumptions about the `sighash`, as the preimage is not known to the signer. */
```
Comment (sdaftuar) [inline src/script/interpreter.h]: This should emphasize that the issue about not knowing the hash preimage is for the verifier, not the signer.  Perhaps:
```
/** A low level signature checker.  Note that the caller must verify that the hash passed in relates to a known message (unlike for the other signature checkers). */

**Review by sdaftuar — COMMENTED** [review]
@kallewoof Thanks for providing that context, now I understand a bit better why this is being implemented in this way.  My only suggestion is that we improve the comments to explain the implications of this for verification.

File: src/script/interpreter.h (hunk 1 above)
Comment (kallewoof) [reply src/script/interpreter.h]: Sorry for delay. Yes, that looks great, thanks. Adopting.

**Comment by jtimon** [comment]
utACK 0bd274fe8544c896616268a21edd06b2d927fc0f

An alternative that perhaps @sdaftuar would like more given his reasonings, is having a templated class that will work for any object implementing a GetHash() method.
Something like:

```
explicit SimpleSignatureChecker(const T& hashable) : hash(hashable.GetHash()) {}
```
But that would require a wrapper class for signet and another one for the signed messages, I think.
And if a comment solves the concerns, that's certainly easier. Just thinking out loud.

File: src/script/interpreter.h [hunk 2]
```diff
@@ -161,6 +161,18 @@ class BaseSignatureChecker
...
 
+/** A low level signature checker.  Note that the caller must verify that the hash passed in relates to a known message (unlike for the other signature checkers). */
+class SimpleSignatureChecker : public BaseSignatureChecker
+{
+private:
+    uint256 hash;
+
+public:
+    const uint256& GetHash() const { return hash; }
+    explicit SimpleSignatureChecker(const uint256& hash_in) : hash(hash_in) {}
```
Comment (theStack) [inline src/script/interpreter.h]: nit: Move constructor one line up, immediately after `public` (like for the other class `SimpleSignatureCreator` and for most classes in general I'd say)

File: src/script/sign.h [hunk 3]
```diff
@@ -31,6 +31,17 @@ class BaseSignatureCreator {
     virtual bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& keyid, const CScript& scriptCode, SigVersion sigversion) const =0;
 };
 
+/** A general purpose signature creator. Note that this lacks assumptions about the `sighash`, as the preimage is not known to the signer. */
+class SimpleSignatureCreator : public BaseSignatureCreator
+{
+    SimpleSignatureChecker checker;
+
+public:
+    explicit SimpleSignatureCreator(const uint256& hashIn) : BaseSignatureCreator(), checker(hashIn) {};
```
Comment (theStack) [inline src/script/sign.h]: Is there any reason why the constructor of this class calls its base class constructor `BaseSignatureCreator()`, while the constructor of `SimpleSignatureChecker()` doesn't?

File: src/test/simplesigner_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+#include <tinyformat.h>
+#include <uint256.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_FIXTURE_TEST_SUITE(simplesigner_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(simplesigner_base_tests)
+{
+    uint256 message;
```
Comment (theStack) [inline src/test/simplesigner_tests.cpp]: Is it intended that here (and in all following testcases) `message` is (implicitely) set to all-zero?

**Review by theStack — COMMENTED** [review]
Concept ACK

File: src/script/sign.h (hunk 3 above)
Comment (kallewoof) [reply src/script/sign.h]: Nope. Shouldn't change anything in practice, but consistency is nice. Unifying.

File: src/script/interpreter.h (hunk 2 above)
Comment (kallewoof) [reply src/script/interpreter.h]: I think it's fine as is, and there are plenty of examples of the constructor not being right below `public:`.

File: src/test/simplesigner_tests.cpp (hunk 4 above)
Comment (kallewoof) [reply src/test/simplesigner_tests.cpp]: It shouldn't really matter what the message is, but then there's that saying about bridges, so I'll set it to something, sure.

**Comment by instagibbs** [comment]
concept and approach ACK. It's ok sometimes to have really small tooling to support not-yet-merged uses, provided there are tests

**Comment by kallewoof** [comment]
Closing this as it's no longer certain whether it will be used, now that BIP-322 switched to a tx based version.

**Pull request closed without merging**